### PR TITLE
eclipses: 2022-12 -> 2023-03

### PR DESCRIPTION
**Pull request by jerith666**
* generalize the justj-stripping regex a tad, since it has grown a trailing /java in some flavors of eclipse.

* resurrect .eclipseproduct workaround from 2012-12 / 8a7561ec09957e813048aba119f7459aa392f22d

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @rycee @eliasvasylenko @OPNA2608 

last update was #210057

### CODE REVIEW

**Comment by jerith666** [comment]
marking draft for two reasons:

1. I'm unable to test on `aarch64`
2. I haven't had a chance to use this at work for any length of time; that'll come next week

**Comment by jerith666** [comment]
Marking ready for review; this has been working fine for me at work this week.

On the subject of `aarch64`, I know the hashes are correct, but I'm unable to build the packages because various upstream dependencies don't cross-compile cleanly.  I'll leave it to others to decide on the importance of testing on that platform before merging this.

File: pkgs/applications/editors/eclipse/build-eclipse.nix [hunk 1]
```diff
@@ -54,7 +53,7 @@ stdenv.mkDerivation rec {
     ln -s $out/eclipse/icon.xpm $out/share/pixmaps/eclipse.xpm
 
     # ensure eclipse.ini does not try to use a justj jvm, as those aren't compatible with nix
-    ${perl}/bin/perl -i -p0e 's|-vm\nplugins/org.eclipse.justj.*/jre/bin\n||' $out/eclipse/eclipse.ini
+    ${perl}/bin/perl -i -p0e 's|-vm\nplugins/org.eclipse.justj.*/jre/bin.*\n||' $out/eclipse/eclipse.ini
```
Comment (SuperSandro2000) [inline pkgs/applications/editors/eclipse/build-eclipse.nix:56]:
```suggestion
    perl -i -p0e 's|-vm\nplugins/org.eclipse.justj.*/jre/bin.*\n||' $out/eclipse/eclipse.ini
```
this should be in nativeBuildInputs

Comment (jerith666) [reply pkgs/applications/editors/eclipse/build-eclipse.nix]: remove `${perl}/bin/` and added `perl` to `nativeBuildInputs`.

**Pull request merged**